### PR TITLE
Script to generate IntelliJ Run Configuration for Controller and Invoker

### PR DESCRIPTION
**Pull request by chetanmeh**
Script which generates the run configuration for Controller and Invoker to enable launching them from within IDE. This allows faster development and debugging as time spent in rebuilding and restarting docker containers can be avoided

The script can be run via below command (requires OpenWhisk running)

    ./gradlew -p tools/dev intellij

intellij-run

The script also supports customizing the generated config with local overrides via custom config file (see README.md for details)

For now using this script Controller can be launched from within the IDE. Invoker posses some [problem](apache/incubator-openwhisk#3195)

### CODE REVIEW

**Comment by chetanmeh** [comment]
Can this PR be reviewed?

File: tools/dev/README.md [hunk 1]
```diff
@@ -41,3 +41,71 @@ Sample output
     Processing whisks_design_document_for_activations_db_filters_v2.1.0.json
             - whisks-filters.v2.1.0-activations.js
     Generated view json files in /path/too/tools/build/views
+
+## IntelliJ Run Config Generator
+
+This script enables creation of [Intellij Launch Configuration][1] in _<openwhisk home>/.idea/runConfigurations_ 
+with name controller0 and invoker0. For this to work your Intellij project should be [directory based][3]. If your
```
Comment (rabbah) [inline tools/dev/README.md]: j/J

File: tools/dev/README.md [hunk 2]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+and 'invoker0'
+
+Key points to note
+
+1. Uses ~/tmp/openwhisk/controller (or invoker) as working directory
+2. Changes the PORT to linked one. So controller gets started at 10001 only just like container
+
+Now the docker container can be stopped and application can be launched from within the IDE
+
+**Note** - Currently only controller can be run from IDE. Invoker posses some [problems][2]
```
Comment (rabbah) [inline tools/dev/README.md]: You use the running controller container to capture the env settings?

Comment (chetanmeh) [reply tools/dev/README.md]: Yes

**Comment by chetanmeh** [comment]
@rabbah Can you review this

File: tools/dev/README.md [hunk 3]
```diff
@@ -41,3 +41,71 @@ Sample output
     Processing whisks_design_document_for_activations_db_filters_v2.1.0.json
             - whisks-filters.v2.1.0-activations.js
     Generated view json files in /path/too/tools/build/views
+
+## IntelliJ Run Config Generator
+
+This script enables creation of [Intellij Launch Configuration][1] in _<openwhisk home>/.idea/runConfigurations_ 
+with name controller0 and invoker0. For this to work your Intellij project should be [directory based][3]. If your 
+project is file based (uses ipr files) then you can convert it to directory based via _File -> Save as Directory-Based Format_
```
Comment (rabbah) [inline tools/dev/README.md]: do you mind ending sentences with `.`

File: tools/dev/README.md [hunk 4]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+
+### Usage
+
+First setup OpenWhisk so that Controller and Invoker containers are up and running. Then run the script
+
+    ./gradlew -p tools/dev intellij
+    
+It would inspect the running docker containers and then generate the launch configs with name 'controller0'
```
Comment (rabbah) [inline tools/dev/README.md]: `.`

File: tools/dev/README.md [hunk 5]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+
+## IntelliJ Run Config Generator
+
+This script enables creation of [Intellij Launch Configuration][1] in _<openwhisk home>/.idea/runConfigurations_ 
+with name controller0 and invoker0. For this to work your Intellij project should be [directory based][3]. If your 
+project is file based (uses ipr files) then you can convert it to directory based via _File -> Save as Directory-Based Format_
+
+### Usage
+
+First setup OpenWhisk so that Controller and Invoker containers are up and running. Then run the script
```
Comment (rabbah) [inline tools/dev/README.md]: `:`

File: tools/dev/README.md [hunk 6]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+
+    ./gradlew -p tools/dev intellij
+    
+It would inspect the running docker containers and then generate the launch configs with name 'controller0' 
+and 'invoker0'
+
+Key points to note
```
Comment (rabbah) [inline tools/dev/README.md]: `:`

File: tools/dev/README.md [hunk 7]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+
+1. Uses ~/tmp/openwhisk/controller (or invoker) as working directory
```
Comment (rabbah) [inline tools/dev/README.md]: `,`

File: tools/dev/README.md [hunk 8]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+2. Changes the PORT to linked one. So controller gets started at 10001 only just like container
```
Comment (rabbah) [inline tools/dev/README.md]: `.`

File: tools/dev/README.md [hunk 9]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+
+Now the docker container can be stopped and application can be launched from within the IDE
```
Comment (rabbah) [inline tools/dev/README.md]: `.`

File: tools/dev/README.md (hunk 2 above)
Comment (rabbah) [reply tools/dev/README.md]: s/only controller/only the controller
and `.` at end of sentence 🙏

File: tools/dev/README.md [hunk 10]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+
+**Note** - Currently only controller can be run from IDE. Invoker posses some [problems][2]
+
+### Configuration
+
+The script allows some local customization of the launch configuration. This can be done by creating a [config][4] file
+`intellij-run-config.groovy` in project root directory. Below is an example of _<openwhisk home>/intellij-run-config.groovy_ 
+file to customize the logging and db port used for CouchDB
```
Comment (rabbah) [inline tools/dev/README.md]: `.`

File: tools/dev/README.md [hunk 11]
```diff
@@ -41,3 +41,71 @@ Sample output
...
+            'logback.configurationFile':'/path/to/custom/logback.xml'
+    ]
+    env = [
+            'DB_PORT' : '5989'
+    ]
+}
+
+```
+
+The config allows following properties
```
Comment (rabbah) [inline tools/dev/README.md]: `:`

File: tools/dev/README.md (hunk 7 above)
Comment (rabbah) [reply tools/dev/README.md]: why you use system temp directory? anyway, minor issue... worked fine for me.

File: tools/dev/README.md (hunk 1 above)
Comment (rabbah) [reply tools/dev/README.md]: can you add a tip: how do you use the run configurations that are generated?

**Comment by chetanmeh** [comment]
Updated the readme based on review feedback. Future PR would not miss on the dots ... :dart:

**Pull request merged**